pytorch_multi_things_2.py

```diff
@@ -0,0 +1,221 @@
+"""
+It seems the hardest part of supporting the API I proposed last week was dealing with amp models.
+This proposal extends the previously proposed TrialContext API with an amp_init() module:
+
+    model = context.model(user_created_model())
+    opt = context.optimizer(user_created_opt(model))
+
+    # This is not ideal, but I think its not the end of the world.
+    model, opt = context.amp_init(model, opt)
+
+I found out that pytorch 1.6 will have its own amp module, and it's API is quite a bit cleaner than
+the apex.amp module (and it seems to use no gloabls :)
+
+It turns out that the pytorch-native amp API is pretty close to the API I proposed already, so I
+made some minor adjustemnts to make sure that they were even closer (and that we could easily
+support pytorch-native AMP in the future).
+
+The question I still have is: can we automatically handle optimizer.zero_grad()?  Is that too
+strong of an assumption?
+"""
+
+############
+
+"""
+Example 1: pytorch-native gradient scaling.
+
+Notice that there is only one GradScaler in the training loop, just like there is only one
+TrialContext object in a Determined PyTorchTrial.
+
+code from https://pytorch.org/docs/master/notes/amp_examples.html#typical-mixed-precision-training
+"""
+
+# Creates model and optimizer in default precision
+model = Net().cuda()
+optimizer = optim.SGD(model.parameters(), ...)
+
+# Creates a GradScaler once at the beginning of training.
+scaler = GradScaler()
+
+for epoch in epochs:
+    for input, target in data:
+        optimizer.zero_grad()
+
+        # Runs the forward pass with autocasting.
+        with autocast():
+            output = model(input)
+            loss = loss_fn(output, target)
+
+        # Scales loss.  Calls backward() on scaled loss to create scaled gradients.
+        # Backward passes under autocast are not recommended.
+        # Backward ops run in the same precision that autocast used for corresponding forward ops.
+        scaler.scale(loss).backward()
+
+        # scaler.step() first unscales the gradients of the optimizer's assigned params.
+        # If these gradients do not contain infs or NaNs, optimizer.step() is then called,
+        # otherwise, optimizer.step() is skipped.
+        scaler.step(optimizer)
+
+        # Updates the scale for next iteration.
+        scaler.update()
+
+
+# The same thing in Determined:
+
+class MyPytorchTrial(PyTorchTrial):
+    def __init__(self, context):
+        # There's one context, just like there was one GradScaler()
+        # our context object will mimic the API of the GradScaler.
+        self.context = context
+
+        self.scaler = self.context.set_scaler(GradScaler())
```

Comment:
So it seems set_scaler() takes the place of amp_init()? There's nothing else that needs to be done (other than some safety checks similar to what we do on configuring Apex today, and setting _use_amp, etc. - and that can all just be done in set_scaler().

Reply:
I think the pytorch API is quite a bit more flexible than Apex (right?) in which case I'd rather we stick to the pytorch API, personally.  @katport has dealt with converting a pytorch amp model to apex amp, perhaps she has thoughts?